qiskit_experiments/data_processing/nodes.py

```diff
@@ -56,130 +106,266 @@ def _format_data(self, datum: Any) -> Any:
+                angle = np.arctan(self._main_axes[idx][1] / self._main_axes[idx][0])
+                error_value = np.sqrt(
+                    (error[idx][0] * np.cos(angle)) ** 2 + (error[idx][1] * np.sin(angle)) ** 2
+                )
+                processed_error.append(error_value)
+
+        return np.array(processed_data), processed_error
+
+    def train(self, data: List[Any]):
+        """Train the SVD on the given data.
+
+        Each element of the given data will be converted to a 2D array of dimension
+        n_qubits x 2. The number of qubits is inferred from the shape of the data.
+        For each qubit the data is collected into an array of shape 2 x n_data_points.
+        The mean of the in-phase a quadratures is subtracted before passing the data
+        to numpy's svd function. The dominant axis and the scale is saved for each
+        qubit so that future data points can be projected onto the axis.
+
+        Args:
+            data: A list of datums. Each datum will be converted to a 2D array.
+        """
+        if not data:
+            return
+
+        n_qubits = self._format_data(data[0])[0].shape[0]
+
+        self._main_axes = []
+        self._scales = []
+        self._means = []
+
+        for qubit_idx in range(n_qubits):
+            datums = np.vstack([self._format_data(datum)[0][qubit_idx] for datum in data]).T
+
+            # Calculate the mean of the data to recenter it in the IQ plane.
+            mean_i = np.average(datums[0, :])
+            mean_q = np.average(datums[1, :])
+
+            self._means.append((mean_i, mean_q))
+
+            datums[0, :] = datums[0, :] - mean_i
+            datums[1, :] = datums[1, :] - mean_q
+
+            mat_u, mat_s, _ = np.linalg.svd(datums)
+
+            self._main_axes.append(mat_u[:, 0])
+            self._scales.append(mat_s[0])
 
-    def _process(self, datum: np.array) -> np.array:
-        """Take the imaginary part of the IQ data.
+
+class ToReal(IQPart):
+    """IQ data post-processing. Isolate the real part of single-shot IQ data."""
+
+    def _required_dimension(self) -> Set[int]:
+        """Require memory to be a 2D or a 3D array."""
+        return {2, 3}
+
+    def _process(
+        self, datum: np.array, error: Optional[np.array] = None, **options
+    ) -> Tuple[np.array, np.array]:
+        """Take the real part of the IQ data.
 
         Args:
-            datum: A 3D array of shots, qubits, and a complex IQ point as [real, imaginary].
+            datum: A 2D or 3D array of shots, qubits, and a complex IQ point as [real, imaginary].
+            error: An optional 2D or 3D array of shots, qubits, and an error on a complex IQ point
+                as [real, imaginary].
 
         Returns:
-            A 2D array of shots, qubits. Each entry is the imaginary part of the given IQ data.
+            A 1D or 2D array, each entry is the real part of the given IQ data and error.
         """
         if self.scale is None:
```

Comment:
if `self.scale` returns 1.0 when not assigned, we can remove this branching.

Reply:
Updated in 935afda